trin-core/src/types/accumulator.rs

```diff
@@ -0,0 +1,214 @@
+use ethereum_types::U256;
+use ssz::Encode;
+use ssz_derive::{Decode, Encode};
+use tree_hash::{MerkleHasher, TreeHash};
+use tree_hash_derive::TreeHash;
+
+use crate::types::header::Header;
+
+const EPOCH_SIZE: u32 = 8192; // number of blocks / epoch
+const MAX_HISTORICAL_EPOCHS: u32 = 131072; // 2 ** 17
+
+// ssz container
+#[derive(Clone, Debug, Decode, Encode, TreeHash)]
+pub struct MasterAccumulator {
+    historical_epochs: HistoricalEpochs,
+    current_epoch: EpochAccumulator,
+}
+
+// todo: remove dead-code exception
+#[allow(dead_code)]
+impl MasterAccumulator {
+    fn new() -> Self {
+        Self {
+            historical_epochs: HistoricalEpochs { epochs: vec![] },
+            current_epoch: EpochAccumulator {
+                header_records: vec![],
+            },
+        }
+    }
+
+    // as defined in: https://github.com/ethereum/portal-network-specs/blob/master/history-network.md
+    fn update_accumulator(&mut self, new_block_header: &Header) {
+        // get the previous total difficulty
+        let last_total_difficulty = match self.current_epoch.header_records.len() {
+            // genesis
+            0 => U256::from_dec_str("0").unwrap(),
+            _ => {
+                self.current_epoch
+                    .header_records
+                    .last()
+                    .unwrap()
+                    .total_difficulty
+            }
+        };
+
+        // check if the epoch accumulator is full
+        if self.current_epoch.header_records.len() == EPOCH_SIZE as usize {
+            // compute the final hash for this epoch
+            let epoch_hash = self.current_epoch.tree_hash_root();
+            // append the hash for this epoch to the list of historical epochs
+            self.historical_epochs.epochs.push(epoch_hash);
+            // initialize a new empty epoch
+            self.current_epoch = EpochAccumulator {
+                header_records: vec![],
+            };
+        }
+
+        // construct the concise record for the new header and add it to the current epoch
+        let header_record = HeaderRecord {
+            block_hash: new_block_header.hash(),
+            total_difficulty: last_total_difficulty + new_block_header.difficulty,
+        };
+        self.current_epoch.header_records.push(header_record);
+    }
+}
+
+// ssz list
+// max_length = MAX_HISTORICAL_EPOCHS
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct HistoricalEpochs {
+    epochs: Vec<tree_hash::Hash256>,
+}
```

Comment:
Can we use [`VariableList`](https://docs.rs/eth2_ssz_types/0.2.1/ssz_types/struct.VariableList.html) to enforce max length?